### ISSUE TITLE
WARNING: Unable to load session

### ISSUE THREAD

**Issue by iandshaw** [aws/aws-dynamodb-session-tomcat#23]
Hi,

Using the 2.0.0 release and when I have more than 1 instance in the auto scaling group and sticky sessions disabled I see the following errors below catalina.log with various different session ids.  If I use the explore table option in aws console it takes a while for the session ids in the log file to be visible.

I'm just using the default read/write capacity and the cloudwatch metrics are barely above zero so they appear to be sufficient.  

Also strangely the createdAt and updatedAt values are always 0.  

Apr 12, 2015 5:12:53 AM com.amazonaws.tomcatsessionmanager.amazonaws.services.dynamodb.sessionmanager.DynamoDBSessionStore load
WARNING: Unable to load session with id 7E4251C5084B7DCFCA3CC56F1F368EFF

Any ideas?

Thanks
Ian

**Comment by shorea** [comment]
Hey Ian. Are you upgrading from a version prior to 2.0.0 or is this a fresh install of the session manager?

**Comment by iandshaw** [comment]
Hi, it is fresh install. Thanks

**Comment by shorea** [comment]
As for why the session ids take a while to be visible in the table, only idle sessions are persisted to DynamoDB. Active sessions are kept in memory. After 30 seconds a session becomes idle and is eligible to be backed up in DynamoDB. At that time it may exist it both memory and be backed up in Dynamo or may be swapped out completely and only exist in Dynamo. The properties maxIdleBackup, maxIdleSwap, and minIdleSwap influence this behavior (https://tomcat.apache.org/tomcat-4.0-doc/config/manager.html).

The createdAt and updatedAt attributes are no longer used in the 2.0 version and exist only to preserve backwards compatibility with 1.x versions. They will be removed in a later version so they no longer appear in the table.

The warnings you are seeing appear to be due to your non-sticky session configuration. Since sessions may exist but may not be in the backing Dynamo table (i.e. active sessions that exist only in memory), other servers will try to load the session but find it missing from both their in memory session storage and in Dynamo. Because of this non-sticky sessions are not supported with this session manager (nor the JDBC or File session manager bundled with Tomcat).

In summary the session manager for DynamoDB is primarily a backup store for idle sessions and persistent storage of sessions across restarts of Tomcat and is not a real time representation of your active sessions across your clustered environment. This is not made clear in the documentation, I'll make sure that this is stated explicitly in the ReadMe and in the getting started guide.

**Comment by iandshaw** [comment]
Certainly makes more sense now I had assumed that sticky sessions were not required.

I think it was this statement in the limitations section of the documentation that lead me to the assumption that they were not required.  If the sessions are only backed up after an idle time how do these limitations become apparent?:

Limitations (http://docs.aws.amazon.com/AWSSdkDocsJava/latest//DeveloperGuide/java-dg-tomcat-session-manager.html)
The session manager does not support session locking. Therefore, applications that use many concurrent AJAX calls to manipulate session data may not be appropriate for use with the session manager, due to race conditions on session data writes and saves back to the data store.

**Comment by shorea** [comment]
Sorry for the delay in response @iandshaw. The limitations documentation is definitely confusing and borderline incorrect. You are correct, the writes back to Dynamo would not cause the session manager to be inappropriate for concurrent session access. With sticky sessions all modifications to session data will be handled by the same server and thus only threads running in the same JVM will be concurrently modifying session data which can be resolved by synchronizing on the session object when writing to it.

I'll add this excerpt to the documentation cleanup task to make this less confusing.

**Comment by iandshaw** [comment]
thanks mate
